### PR TITLE
Pr86

### LINKED ISSUE #86: Generalize signals
# Issue
## Problem Description
The signal.py file assumes that data can be represented as a numpy array. This is not true for all use-cases and we should generalize this to accept an index and the handler instead to fetch the data.

## Expected Behavior
Ability to evaluate signals for arbitrary datasets.

## What Needs to be Done
Update all the functions in signal.py to accept handler and indices.

## How Can It Be Tested or Reproduced
Run attacks.

### LINKED ISSUE #87: Skip indices in LiRA
# Issue
## Problem Description
When computing skip indices for LiRA, nothing is done with the identified data points. We should follow the RMIA approach and update the relevant quantities. Furthermore, this eliminates the need to use nanmin and nanmax.

## Expected Behavior
Data points without an in or an out model should not be considered for online auditing.

## What Needs to be Done
- Replicate RMIA functionality. Perhaps bring it out to shadow model handler? 
- Ensure the subsets created in RMIA  and LiRA does not rely on code from old dataset functions.

## How Can It Be Tested or Reproduced
Run attack and check logger outputs.

### LINKED ISSUE #88: No _labels in datset for RMIA
# Issue
## Problem Description
We are expecting the dataset to contain a field _labels for RMIA (and possibly other attacks). Datasets do not necessarily have this field.

## Expected Behavior
RMIA should run without the _label field.

## What Needs to be Done
Implement a general way to extract the labels from a dataloader.

## How Can It Be Tested or Reproduced
Run attack.

### PR DESCRIPTION
**Pull request by johanos1**
# Description

Summary of changes

* Generalized and cleaned signal.py. Instead of accepting a dataset, it now accepts the handler and the relevant indices
* Handled LiRA skip indices similarly to RMIA 
* Added get_labels to input handler. Replaced dependencies of legacy dataloader in rmia.py and qmia.py

## Resolved Issues

- [ ] fixes #88 
- [ ] fixes #87 
- [ ] fixes #86 

## How Has This Been Tested?
All MIA attacks have been run under both online and offline (when applicable) and are passing.

### CODE REVIEW

**Review by fazelehh — APPROVED** [review]
Nice with passing handler instead of the dataset. Makes things neater.
I think it makes the review process better if we focus on one feature or bug per PR.

File: leakpro/attacks/mia_attacks/lira.py [hunk 1]
```diff
@@ -103,45 +103,46 @@ def prepare_attack(self:Self)->None:
...
-        self.audit_data = self.get_dataloader(self.audit_dataset["data"]).dataset
+        self.in_indices_mask = ShadowModelHandler().get_in_indices_mask(self.shadow_model_indices, self.audit_dataset["data"]).T
 
         # Check offline attack for possible IN- sample(s)
-        if not self.online:
+        if self.online:
+            # filter out the points that no shadow model has seen and points that all shadow models have seen
+            num_shadow_models_seen_points = np.sum(self.in_indices_mask, axis=0)
+            # make sure that the audit points are included in the shadow model training (but not all)
+            mask = (num_shadow_models_seen_points > 0) & (num_shadow_models_seen_points < self.num_shadow_models)
```
[the remark this answers is not in the document]
Comment (johanos1) [reply leakpro/attacks/mia_attacks/lira.py:113]: I think that something we can discuss but then the follow-up question is: given n models, how many in/out models should we require? This code only ensures that the audit dataset is actually auditable.

File: config/audit.yaml [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
-      training_data_fraction: 0.5  # Fraction of the auxilary dataset (data without train and test indices) to use for training the quantile regressor
-      epochs: 5  # Number of training epochs for quantile regression
-    population:
-      attack_data_fraction: 0.1  # Fraction of the auxilary dataset to use for this attack
-    lira:
-      training_data_fraction: 0.1  # Fraction of the auxilary dataset to use for this attack (in each shadow model training)
-      num_shadow_models: 8 # Number of shadow models to train
-      online: False # perform online or offline attack
-      fixed_variance: True # Use a fixed variance for the whole audit
+    # rmia:
```
[the remark this answers is not in the document]
Comment (johanos1) [reply config/audit.yaml:4]: Mm, maybe we should stop pushing this file altogether?

[the remark this answers is not in the document]
Comment (johanos1) [reply config/audit.yaml:4]: If so, let's add it to gitignore in a separate PR

[the remark this answers is not in the document]
Comment (fazelehh) [reply config/audit.yaml:4]: sometimes it is needed like when adding new attacks. Or when adding FL this file might be changed.

File: leakpro/attacks/mia_attacks/lira.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fazelehh) [reply leakpro/attacks/mia_attacks/lira.py:113]: I can answer this question with two different objectives: 
1. Maximizing the Number of Auditable Data Points: This is essentially what we are doing now. If a data point has at least one shadow model, we audit it. 

2. Ensuring Reliable Audit Points: We can provide recommendations to the user for more reliable audits. For instance, if the ratio of data points for each shadow model and the number of shadow models results in fewer than (let's say) 4 shadow models per data point, we can warn the user. In the first principle paper, good results were observed with around eight in and eight out shadow models per data. Alternatively, we can calculate the mean number of shadow models per audit point and filter out points below this mean or the first quartile. Whatever we do here might be model, and data-dependent.

[the remark this answers is not in the document]
Comment (johanos1) [reply leakpro/attacks/mia_attacks/lira.py:113]: Yet another approach could be to assign some confidence based on the in/out numbers

[the remark this answers is not in the document]
Comment (johanos1) [reply leakpro/attacks/mia_attacks/lira.py:113]: In the end, we will essentially have a stronger attack mounted on some of the data points than others. Then, if we do the attack over a few consecutive rounds, we should be good?

[the remark this answers is not in the document]
Comment (johanos1) [reply leakpro/attacks/mia_attacks/lira.py:113]: I think this discussion belongs to issue 79, Ill paste it there.

**Pull request merged**